android-pdf-viewer/src/main/java/com/github/barteksc/pdfviewer/annotation/core/pdf/PdfUtil.kt

```diff
@@ -0,0 +1,327 @@
+                                annotationsList.add(extractedAnnotation)
+                            }
+                        }
+                    } else {
+                        logError(TAG, "Annotation is not recognised")
+                    }
+                }
+            }
+            return annotationsList
+        } catch (e: Exception) {
+            e.message?.let { logError(TAG, it) }
+            e.printStackTrace()
+            return emptyList()
+        }
+    }
+
+    private fun getExtractedSquareAnnotation(
+        annotation: PdfDictionary,
+        llx: Float,
+        lly: Float,
+        urx: Float,
+        ury: Float
+    ): Annotation {
+        val xBottomLeftPoint = llx
+        val yBottomLeftPoint = lly
+        val bottomLeftPoint = PointF(xBottomLeftPoint, yBottomLeftPoint)
+        val xTopRightPoint = urx
+        val yTopRightPoint = ury
+        val topRightPoint = PointF(xTopRightPoint, yTopRightPoint)
+
+        // from the extracted coordinates, calculate the rest
+        val squareAnnotationPoints =
+            generateRectangleCoordinates(bottomLeftPoint, topRightPoint)
+
+        val relationsArray: PdfArray? =
+            annotation.getAsArray(PdfName("relations"))
+
+        return Annotation(
+            AnnotationType.SQUARE.name,
+            squareAnnotationPoints,
+            relations = getExtractedRelations(relationsArray)
+        )
+    }
+
+    private fun getExtractedRelations(relationsArray: PdfArray?): Relations? {
+        val documentations = mutableListOf<Documentation>()
+        if (relationsArray != null) {
+            for (j in 0 until relationsArray.size()) {
+                val documentationDict: PdfDictionary =
+                    relationsArray.getAsDict(j)
+                val schemaId: PdfNumber? =
+                    documentationDict.getAsNumber(PdfName("schemaId"))
+                val documentId: PdfString? =
+                    documentationDict.getAsString(PdfName("documentId"))
+
+                if (schemaId != null && documentId != null) {
+                    documentations.add(
+                        Documentation(
+                            schemaId.intValue().toLong(),
+                            documentId.toString()
+                        )
+                    )
+                }
+            }
+        } else {
+            return null
+        }
+        return Relations(documentations)
+    }
+
+    /** Map annotations to shapes,
+     *  using the page height when converting between PDF space and image space */
+    private fun getShapesFor(
+        pdfAnnotations: List<Annotation>,
+        pageHeight: Int
+    ): List<Shape> {
+        // convert annotation to shape
+        val shapes = pdfAnnotations.map { annotation ->
+            when (annotation.type) {
+                AnnotationType.SQUARE.name -> return@map annotation.toRectangleShape(pageHeight)
+                else -> throw Exception("Annotation is not recognised")
```

Comment:
Throws exceptions but callers are not caching this. Instead return an empty list

Reply:
Done

Reply:
`Shape.toAnnotation` still throws here - can you handle it?